File: apps/adk-worker/src/fleetscope_worker/scenario.py
"""What a run is allowed to do, decided here and not by the caller.

# Why every knob lives in this file

The request contract carries identifiers only. Target, model, retry budget,
fault policy and call ceiling are all declared here, in server source, because
each of them either spends money or reaches the internet. A caller that could
set `target` could point the worker anywhere; a caller that could set
`maxModelCalls` could spend the budget; a caller that could set `faultAttempts`
could turn the recovery demo into an unbounded retry loop.

These values mirror `LIVE_SCENARIOS` in `packages/run-ledger/src/scenario.ts`.
The API admits a run against that definition and this worker executes it against
this one, so `test_scenario.py` asserts the two agree rather than trusting that
two files in two languages were edited together.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Scenario:
    id: str
    root_agent: str
    delegated_agent: str
    #: The only repository this scenario may read. Not caller-supplied.
    target: str
    #: Default model. A deployment may override it via FLEETSCOPE_ADK_MODEL;
    #: a request may not, because a request is untrusted and a model name is a
    #: cost decision.
    model: str
    max_model_calls: int
    max_retries: int
    #: How many leading attempts the Controlled Fault fails. Fixed, so the demo
    #: is reproducible and cannot be turned into a retry storm.
    fault_attempts: int
    recovery_action: str
    side_effect_class: str
    timeout_s: float


DEPENDENCY_ONBOARDING = Scenario(
    id="dependency_onboarding",
    root_agent="dependency_onboarding",
    delegated_agent="security_review",
    target="google/adk-python",
    model="gemini-2.5-flash",
    max_model_calls=6,
    max_retries=1,
    fault_attempts=1,
    recovery_action="retry_idempotent_read",
    side_effect_class="idempotent_read",
    timeout_s=90.0,
)

SCENARIOS: dict[str, Scenario] = {DEPENDENCY_ONBOARDING.id: DEPENDENCY_ONBOARDING}


class UnknownScenario(Exception):
    """The scenario id is not one this worker implements."""


def find_scenario(scenario_id: str) -> Scenario:
    scenario = SCENARIOS.get(scenario_id)
    if scenario is None:
        raise UnknownScenario(f"{scenario_id!r} is not an allowlisted scenario")
    return scenario

File: packages/run-ledger/src/scenario.ts
/**
 * The scenarios a live run may use. There is no other way to start one.
 *
 * # Why an allowlist and not parameters
 *
 * A live run spends money and reaches the public internet. Every field that
 * could widen either — the prompt, the target, the tool, the retry budget — is
 * fixed HERE, in the server's own source, and the request names a scenario
 * rather than describing one. There is deliberately no endpoint anywhere that
 * accepts a prompt, a URL, a shell command or a tool parameter.
 *
 * This mirrors `apps/api/src/live/allowlist.ts`, which bounds the existing
 * single-decision path the same way.
 */

/** What a recovery is permitted to do. Only one class is implemented. */
export type SideEffectClass = 'idempotent_read';

export interface LiveScenario {
  readonly id: string;
  readonly description: string;
  /** The ADK root agent this scenario runs. */
  readonly rootAgent: string;
  /** The sub-agent the root is expected to delegate to. */
  readonly delegatedAgent: string;
  /** The allowlisted external target. Never taken from a request. */
  readonly target: string;
  /** Human-readable description of the external operation. */
  readonly externalOperation: string;
  /** The single recovery the Warden policy may authorize. */
  readonly recoveryAction: 'retry_idempotent_read';
  readonly sideEffectClass: SideEffectClass;
  /** At most one Warden retry, by construction. */
  readonly maxWardenRetries: 1;
  readonly maxModelCalls: number;
  readonly maxOutputTokens: number;
  readonly timeoutMs: number;
}

export const LIVE_SCENARIOS: readonly LiveScenario[] = [
  {
    id: 'dependency_onboarding',
    description:
      'A root agent onboards a dependency and delegates a security review that reads public repository metadata.',
    rootAgent: 'dependency_onboarding',
    delegatedAgent: 'security_review',
    // A public, read-only target. No purchase, activation, email, account
    // mutation or write of any kind is reachable from this scenario.
    target: 'google/adk-python',
    externalOperation: 'read-only repository metadata',
    recoveryAction: 'retry_idempotent_read',
    sideEffectClass: 'idempotent_read',
    maxWardenRetries: 1,
    maxModelCalls: 6,
    maxOutputTokens: 256,
    timeoutMs: 90_000,
  },
];

export function findScenario(id: string): LiveScenario | null {
  return LIVE_SCENARIOS.find((scenario) => scenario.id === id) ?? null;
}
